### ISSUE TITLE
auth: Fix sign out for proxy auth providers

### ISSUE THREAD

**Issue by Rugvip** [backstage/backstage#12016]
## Expected Behavior

For any of the [proxied auth providers](https://backstage.io/docs/auth/#sign-in-with-proxy-providers) the logout should either just work, link to the logout page, or be disabled.

## Actual Behavior

The sign out button does nothing.

## Steps to Reproduce

1. Set up any of the proxied auth providers and sign in.
2. Head to the user settings page and select "Sign Out" in the kebab menu on the profile card.

## Context

#11954

**Comment by antoniordz96** [comment]
We are using the [oauth2-proxy](https://oauth2-proxy.github.io/oauth2-proxy/) container for authentication. Seems like the container exposes an endpoint `/oauth2/sign_out - this URL is used to clear the session cookie`. Is it possible in the current implementation of backstage to set this button to hit this URL?

**Comment by Rugvip** [comment]
I don't think the logout URL is oauth2 proxy is something that's possible to configure, so no need to make it configurable on our end either tbh. All we really want to do is to make sure the page navigates there when the logout button is clicked.

**Comment by justinjao** [comment]
Hi @Rugvip (and others!) - is this issue still free to work on? If so, I'd like to take a crack at it!

I'm Justin! I'm new to Backstage, and open source in general, but would love to try my hand at contributing!

**Comment by jhaals** [comment]
Hi @justinjao, you are very welcome to give it a go 👏

**Comment by justinjao** [comment]
Hi @jhaals - thanks for getting back to me! I'll try and give it a shot. Just managed to finish setting it up locally (I think).

Currently trying to read through the documentation regarding `proxy auth providers` above, and how exactly to set this up, so that I can reproduce the issue. I'm also reading through setting it up with [Google IAP](https://backstage.io/docs/auth/google/gcp-iap-auth), and if I understand correctly so far, I'll need to set up my own IAP account for this, so planning on trying that for now.

If I misunderstood anything, do let me know. Otherwise, I'll just carry on for now. 🙂

**Comment by justinjao** [comment]
Sorry I haven't followed up - got caught up with other commitments (namely university exams!). 

If it's still free, I'm still very much interested in contributing. I'm hopeful I'll have some time next week to revisit it (I have a bit of a winter break, after exams).

**Comment by loubevi** [comment]
Not sure if this is a separate issue or not, but there is a second way to sign out in Settings->Authentication Providers-> "Sign Out" button.  This seems to have a different behavior than the kebab menu on the Settings->General->Profile card, which does seem to be working for me (Okta).  

The Sign Out button in the Authentication Providers tab seems to be trying to send a refresh token to Okta and I am not sure if that is the correct behavior or not.

**Comment by sross-tc** [comment]
Any news on this?

**Comment by spark72018** [comment]
Is this still an issue that can be picked up? Looking to be a first time contributor 😄

**Comment by benjdlambert** [comment]
@spark72018 it's all yours! I don't think any progress has been made on this yet!

**Comment by justinjao** [comment]
@spark72018 Yes, please feel free to work on it! I've been caught up with other responsibilities, and realistically won't be able to work on it anytime soon. Thank you for picking it up, and good luck!

**Comment by freben** [comment]
Assigned!

**Comment by npiyush97** [comment]
@Rugvip 
how about this for a solution that only redirects to the signing page if the user logs out of signinpage provider?
like if a user is logged in with Google and GitHub but the sign-in provider is Github, and the user sign out of Google does nothing but if the user logs out of GitHub then logs out and redirects to sign-in page.

**Comment by Rugvip** [comment]
@npiyush97 that's a slightly different issue but it also makes sense, let's discuss it in the PR. This issue is about actually performing the logout action when using a proxies auth provider, which requires navigation to a particular logout URL provided by the proxy.

**Comment by koshrf** [comment]
Trying to setup backstage with oauth2-proxy connected to a keycloak, I'm able to login without problems, but the sign out button on the profile does nothing at all and just refresh the page.

This is the log I get from backstage when clicking on the sign out button:

```sh
{"level":"info","message":"::1 - - [13/Feb/2024:17:12:21 +0000] \"GET / HTTP/1.1\" 200 - \"-\" \"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0
.0 Safari/537.36\"","service":"backstage","type":"incomingRequest"}                                                                                                                   {"level":"info","message":"::1 - - [13/Feb/2024:17:12:21 +0000] \"GET /static/vendor.c7c17463.css HTTP/1.1\" 200 - \"-\" \"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, 
like Gecko) Chrome/121.0.0.0 Safari/537.36\"","service":"backstage","type":"incomingRequest"}                                                                                         {"level":"info","message":"::1 - - [13/Feb/2024:17:12:21 +0000] \"GET /favicon.ico HTTP/1.1\" 200 - \"-\" \"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chr
ome/121.0.0.0 Safari/537.36\"","service":"backstage","type":"incomingRequest"}                                                                                                        {"level":"info","message":"::1 - - [13/Feb/2024:17:12:21 +0000] \"GET /api/catalog/entities/by-name/User/default/backstageadmin HTTP/1.1\" 404 237 \"-\" \"node-fetch/1.0 (+https://gi
thub.com/bitinn/node-fetch)\"","service":"backstage","type":"incomingRequest"}                                                                                                        {"component":"token-factory","level":"info","message":"Issuing token for user:default/backstageadmin, with entities user:default/backstageadmin","plugin":"auth","service":"backstage"
,"type":"plugin"}                                                                                                                                                                     {"level":"info","message":"::1 - - [13/Feb/2024:17:12:21 +0000] \"GET /api/auth/oauth2Proxy/refresh HTTP/1.1\" 200 677 \"-\" \"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36\"","service":"backstage","type":"incomingRequest"}
```

I'm doing something wrong ? because I don't see anywhere it sends a logout signal to the oauth2-proxy or clear the tokens, it just ask for a refresh to the proxy.

**Comment by camilaibs** [comment]
Hi @koshrf, this is exactly what this issue is about, thanks for letting us know that this is happening to you as well.
We're always looking for contributors, would you be interested in submitting a pull request 🙂 ?

**Comment by koshrf** [comment]
> Hi @koshrf, this is exactly what this issue is about, thanks for letting us know that this is happening to you as well. We're always looking for contributors, would you be interested in submitting a pull request 🙂 ?

Hi, sorry, I'm not a TS/React/JS developer, I can't fix this kind of bug, I'm on the devops side with an extensive infrastructure that I could use to test if anyone provides a fix to this.

**Comment by npiyush97** [comment]
@koshrf can you test with this #23480

**Comment by Rugvip** [comment]
@npiyush97 could you perhaps create a separate issue for the bug you're aiming to fix? It's not related to this issue

**Comment by npiyush97** [comment]
> @npiyush97 could you perhaps create a separate issue for the bug you're aiming to fix? It's not related to this issue

Yes, i know just remainder to think of this issue.
As you said this does not log the user out of the app just removes user credentials thus user can linger in the app in settings as soon comes out of API,home or catalog network goes mental with refresh request which can be some issue ?.
thinking of how to fix this 
1) As backstage provides multiple auth so if user logs out of their auth-provider (1st auth they logged in with) redirect to the login page as general setting does.
2)create a global state or kind of context that keep track of all auth provider user is logged in with if the state is empty which user not logged in with anything so redirect to login page.
I'm in favour of 1st option. 2nd sound handful if someone wants they can implement themselves <add some emoji>.
Do you have anything other in your mind ?
Now i understand issue 😜?

**Comment by Nikunj0601** [comment]
@Rugvip Okay, got it.

**Comment by ganievs** [comment]
Hi @Rugvip! You could assign the issue on me I'll tackle it on the week.

**Comment by vinaykv23** [comment]
Hi  @freben @Rugvip 
"@backstage/plugin-user-settings": "^0.8.7"
we are using okta sign in and on click of sign out from "General tab" or "Authentication Providers tab" its redirecting to home page, please let me know how to fix this

**Comment by benjdlambert** [comment]
@vinaykv23 I think that actually your issue is something separate to this thinking about it a little bit.

I think that it's to do with the `auto` prop that you're using, which means that when you click signout, you don't have another page to go to as it's going to try and automatically sign you back in again as that's what the `auto` prop does.

I would encourage you to open up a separate issue here, as I think that this is a bit of an unexpected corner case that we didn't think about.

**Comment by mcintoac-aws** [comment]
Any updates on this? Copying from a different ticket:

> On top of the auto functionality not redirecting to the sign-in package again on sign-out, for custom OAuth 2 providers, we require the ability to customize the sign-out functionality. This is necessary for Amazon Cognito, which via the hosted-ui, creates an auth-2 refresh cookie. This cookie requires hitting a Cognito logout API to invalidate. Without the ability to customize the signout functionality for our custom provider, we can not invalidate this cookie, and even without the automatic redirect, clicking the "sign-in" button would automatically sign-in the last user rather than allow a different user sign-in.

https://github.com/backstage/backstage/issues/26440

**Comment by camilaibs** [comment]
^ Is there anything new you can share with us, @craigtracey?

**Comment by gushob21** [comment]
This is still an issue. I am using gcpIap(which is a proxy) and unable to logout either from settings>General>profile or settings>Authentication Providers>Available Providers

**Comment by freben** [comment]
I'll try pinging @craigtracey again! We'd love to have this moved forward.
